test/helpers.js

```diff
@@ -0,0 +1,254 @@
+    buf.writeUint16(typeof opts.version !== 'undefined' ? opts.version : 0x0303);
+    buf.writeUint16(typeof opts.contentLength !== 'undefined' ? opts.contentLength : content.byteLength);
+    buf.writeBytes(content);
+    if (typeof opts.trailer !== 'undefined') {
+      buf.writeBytes(opts.trailer);
+    }
+    return buf.flush();
+  },
+
+  makeEncryptedInnerPlaintext: async function (cipherstate, opts) {
+    const adBuf = new BufferWriter();
+    const innerPlaintextBuf = new BufferWriter();
+    const plaintext = typeof opts.content !== 'undefined' ? opts.content : arrayToBytes([1, 2, 3, 4, 5]);
+    if (typeof opts.innerPlaintext !== 'undefined') {
+      innerPlaintextBuf.writeBytes(opts.innerPlaintext);
+    } else {
+      innerPlaintextBuf.writeBytes(plaintext);
+      innerPlaintextBuf.writeUint8(typeof opts.type !== 'undefined' ? opts.type : 23);
+      if (opts.padding) {
+        innerPlaintextBuf.writeBytes(zeros(opts.padding));
+      }
+    }
+    const ciphertextLength = innerPlaintextBuf.tell() + 16;
+    adBuf.writeUint8(typeof opts.outerType !== 'undefined' ? opts.outerType : 23);
+    adBuf.writeUint16(typeof opts.outerVersion !== 'undefined' ? opts.outerVersion : 0x0303);
+    adBuf.writeUint16(typeof opts.ciphertextLength !== 'undefined' ? opts.ciphertextLength : ciphertextLength);
+    const ciphertext = await cipherstate.encrypt(innerPlaintextBuf.flush(), adBuf.flush());
+    return ciphertext;
+  },
+
+  makeEncryptedRecord: async function (cipherstate, opts) {
+    let ciphertext = opts.ciphertext;
+    if (typeof ciphertext === 'undefined') {
+      ciphertext = await testHelpers.makeEncryptedInnerPlaintext(cipherstate, opts);
+    }
+    return testHelpers.makePlaintextRecord({
+      content: ciphertext,
+      contentLength: opts.outerContentLength,
+      trailer: opts.outerTrailer,
+      type: typeof opts.outerType !== 'undefined' ? opts.outerType : 23,
+      version: opts.outerVersion,
+    });
+  },
+
+  makeCipherState: async function(key, seqnum = 0) {
+    const cipherstate = new CipherState();
+    await cipherstate.setKey(key, ['encrypt', 'decrypt']);
+    cipherstate.seqnum = seqnum;
+    return cipherstate;
+  },
+
+  makeClientHelloRecord: async function(opts, psk) {
+    const clientHello = testHelpers.makeClientHelloMessage(opts);
+    if (psk) {
+      await testHelpers.signClientHelloMessage(clientHello, psk);
+    }
+    return testHelpers.makePlaintextRecord({
+      content: clientHello,
+      type: 22,
+    });
+  },
+
+  makeHandshakeMessage: function (opts) {
+    const buf = new BufferWriter();
+    buf.writeUint8(opts.type || 0);
+    buf.writeVector24(buf => {
+      buf.writeBytes(opts.content || zeros(0));
+    });
+    return buf.flush();
+  },
+
+  makeClientHelloMessage: function(opts) {
+    const buf = new BufferWriter();
+    buf.writeUint8(1);
+    buf.writeVector24(buf => {
+      buf.writeUint16(typeof opts.version !== 'undefined' ? opts.version : 0x0303);
```

Comment:
You use `typeof opts.x !== 'undefined' ? opts.x : blah` here, but the simpler `opts.x || blah` above.  Any particular reason?

Reply:
IIRC the simpler `opts.x || default` form is used when the default is falsy, and hence there's no reason for the caller to want to pass in an explicit falsy value.  Proably cleaner to just use `typeof opts.x !== 'undefined'` consistently throughout.